src/scipp/core/bin_remapping.py

```diff
@@ -1,114 +1,124 @@
-    # We want to write data from all bins along dim to a contiguous chunk in the
-    # content buffer. This will then allow us to create new, larger bins covering the
-    # respective input bins. We use `cumsum` after moving `dim` to the innermost dim.
-    # This will allow us to setup offsets for the new contiguous layout.
-    sizes = var.bins.size()
-    out_dims = [d for d in var.dims if d != dim]
-    out_end = cumsum(sizes.transpose(out_dims + [dim])).transpose(var.dims)
-    out_begin = out_end - sizes
-    out = _cpp._bins_no_validate(
-        data=copy_for_overwrite(var.bins.constituents['data']),
+        return da.data
+
+
+def _with_bin_sizes(var: Variable, sizes: Variable) -> Variable:
+    end = cumsum(sizes)
+    begin = end - sizes
+    data = var if var.bins is None else var.bins.constituents['data']
+    dim = var.dim if var.bins is None else var.bins.constituents['dim']
+    return _cpp._bins_no_validate(data=data, dim=dim, begin=begin, end=end)
+
+
+def _concat_bins(var: Variable, dim: List[str]) -> Variable:
+    # To concat bins, two things need to happen:
+    # 1. Data needs to be written to a contiguous chunk.
+    # 2. New bin begin/end indices need to be setup.
+    # If the dims to concatenate are the *inner* dims a call to `copy()` performs 1.
+    # Otherwise, we first transpose and then `copy()`.
+    # For step 2. we simply sum the (transposed) input bin sizes over the concat dims,
+    # which `_with_bin_sizes` can use to compute new begin/end indices.
+    changed_dims = list(concrete_dims(var, dim))
+    unchanged_dims = [d for d in var.dims if d not in changed_dims]
+    # TODO It would be possible to support a copy=False parameter, to skip the copy if
+    # the copy would not result in any moving or reordering.
+    out = var.transpose(unchanged_dims + changed_dims).copy()
+    sizes = out.bins.size().sum(changed_dims)
+    return _with_bin_sizes(out, sizes)
+
+
+def _combine_bins(var: Variable, coords: Dict[str, Variable], edges: List[Variable],
+                  groups: List[Variable], dim: Dims) -> Dict[str, Variable]:
+    from .binning import make_binned
+    # Overview
+    # --------
+    # The purpose of this code is to combine existing bins, but in a more general
+    # manner than `concat`, which combines all bins along a dimension. Here we operate
+    # more like `groupby`, which combines selected subsets and creates a new output dim.
+    #
+    # Approach
+    # --------
+    # The algorithm works conceptually similar to `_concat_bins`, but with an additional
+    # step, calling `make_binned` for grouping within the erased dims. For the final
+    # output binning, instead of summing the input bin sizes over all erased dims, we
+    # sum only within the groups created by `make_binned`.
+
+    # Preserve subspace dim order of input data, instead of the one given by `dim`
+    changed_dims = [d for d in var.dims if d in concrete_dims(var, dim)]
```

Comment:
`changed_dims` is slightly different here than above (L49), is that intentional?
In addition, it's calling `concrete_dims` multiple times, but I guess that's not an expensive function call?

Reply:
The comment in line 74 explains this. In L49 this does not apply as all changed dims get removed (instead of replaced). Repeated call should not be expensive.